### PR TITLE
add multi lang support using ngx-translate

### PR DESCRIPTION
**Pull request by giuseppetrotta**
I added support for multilingualism using the @ngx-translate library.  

A new module has been added that configures two languages ("en" and "it") and sets english as the default language unless the browser language is set to italian. This module is exported to the shared to make it available throughout the project. 

Each language is stored in a separate .json file. For example the JSON file for the English translation is in assets/i18n/en.json.

This commit include a short translation. E.g.
```json
{
  "navbar": {
  	"group.title": "Group",
	"admin.title": "Admin",
	"lang": {
	  "title": "Language",
	  "en": "English",
      "it": "Italian"
	}
  }
}
```
It is not my intention to provide the complete translation of the core part but if necessary I can give my contribution.  

Translate text using pipe approach as follows:
```html
<a class="nav-link text-nowrap">
  {{"navbar.group.title" | translate}}
</a>
```
Additionally, the multi-lang is enabled using a `MULTI_LANG` env variable to be "injected" using rapydo approach for core variables.

### CODE REVIEW

**Comment by giuseppetrotta** [comment]
One very important thing that I missed. To allow the translation of custom parts it is necessary to be able to create json translation files as an extension of the core ones. This means that they will have to be merged during the build phase (as already happens for the `package.json`)

**Comment by mdantonio** [comment]
Hey, sorry for the long delay, and thank you for the contribution 🙏 

> One very important thing that I missed. To allow the translation of custom parts it is necessary to be able to create json translation files as an extension of the core ones. This means that they will have to be merged during the build phase (as already happens for the package.json)

It makes sense. Json merge is done just before starting UI building, it's a step executed by the docker-entrypoint. You could extend the merge.js script

https://github.com/rapydo/rapydo-angular/blob/3.0/utility/merge.js#L46-L50

In this case I would expect the core json files to be moved elsewhere, and the json files to be served on the asset folder would be those produced by the merge

Would also be needed a parallel PR on the controller to add the env variable being added here

File: src/assets/en.json [hunk 1]
```diff
@@ -0,0 +1,11 @@
+{
```
Comment (mdantonio) [inline src/assets/en.json]: Should these files be under a `i18n` folder?

Or as an alternative the path should be changed accordingly on the translator module
```
return new TranslateHttpLoader(http, "app/rapydo/assets/i18n/", ".json");
```

File: src/app/components/language-translation/language-translation.module.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
+/**
```
Comment (mdantonio) [inline src/app/components/language-translation/language-translation.module.ts]: `components` folder is intended for the components.
modules are on the top level folder (they do not have a proper subfolder), so this file should be just moved 1 level up

File: src/app/components/navbar/navbar.html [hunk 3]
```diff
@@ -81,6 +81,29 @@
            [user]="user"
            (onClick)="isMenuCollapsed = true"
         ></customcontrols>
+        <li class="nav-item dropdown" ngbDropdown *ngIf="multiLang">
+          <a class="nav-link" href="javascript:void(0)" ngbDropdownToggle>
+            <i class="fa fa-language"></i>
+              {{ "navbar.lang.title" | translate }}
```
Comment (mdantonio) [inline src/app/components/navbar/navbar.html]: Maybe the icon itself is enough? Without the label?
On the right side of the navbar others links are all just icons
The Language label could be a tooltip over the icon

File: src/app/components/navbar/navbar.ts [hunk 4]
```diff
@@ -29,6 +30,7 @@ export class NavbarComponent implements OnInit {
 
   public showLogin: boolean = true;
   public allowRegistration: boolean = false;
+  public multiLang: boolean = false;
```
Comment (mdantonio) [inline src/app/components/navbar/navbar.ts]: Better name for this flag could be `enableMultiLang`

File: src/environments/environment.prod.ts [hunk 5]
```diff
@@ -50,6 +51,7 @@ export const environment: Env = {
   forceSSRServerMode:
     base.forceSSRServerMode === "1" || base.forceSSRServerMode === "true",
   spinnerType: base.spinnerType,
+  multiLang: base.multiLang === "1" || base.multiLang === "true",
```
Comment (mdantonio) [inline src/environments/environment.prod.ts]: Better name for this option could be `enableMultiLang` or `multiLangEnabled`
(there is not a great consistency over these options, having both `authEnabled` and `enableFooter`... but in any case `enable` somewhere would be useful

File: utility/config-env.ts [hunk 6]
```diff
@@ -92,7 +92,8 @@ export const environment = {
     allowTermsOfUse: '${process.env.ALLOW_TERMS_OF_USE}',
     minPasswordLength: '${process.env.AUTH_MIN_PASSWORD_LENGTH}',
     forceSSRServerMode: '${process.env.FORCE_SSR_SERVER_MODE}',
-    spinnerType: '${process.env.SPINNER_TYPE}',`;
+    spinnerType: '${process.env.SPINNER_TYPE}',
+    multiLang: '${process.env.MULTI_LANG}',`;
```
Comment (mdantonio) [inline utility/config-env.ts]: Better name for this env variable could be something more explicit, like `ENABLE_MULTI_LANGUAGE`

or even `ENABLE_ANGULAR_MULTI_LANGUAGE` make avoid ambiguity at higher level on project configuration, to clarify that is only intended for angular and not for the backend

**Review by mdantonio — COMMENTED** [review]
A few things to be changed, but overall that's a great starting point, thank you!

extension of merge.js could be added here, or added later on with a followup

File: src/assets/en.json (hunk 1 above)
Comment (giuseppetrotta) [reply src/assets/en.json]: yes sure! I forgot to create that i18n folder

File: src/environments/environment.prod.ts (hunk 5 above)
Comment (giuseppetrotta) [reply src/environments/environment.prod.ts]: the choice is yours! "enable" in the name makes sense

File: utility/config-env.ts (hunk 6 above)
Comment (giuseppetrotta) [reply utility/config-env.ts]: isn't it a config variable only for the front-end UI app? If so, you won't have any disambiguity problems.

File: src/app/components/navbar/navbar.html (hunk 3 above)
Comment (giuseppetrotta) [reply src/app/components/navbar/navbar.html]: yes, it is! I'm not particularly attracted by tooltips, especially if the icon itself is already self-explanatory. I think it is already used for the other icons (eg logout).... then it makes sense

Comment (mdantonio) [reply src/app/components/navbar/navbar.html]: We can eventually skip the tooltip, main point was more about the removal of the label to only keep the icon

File: src/environments/environment.prod.ts (hunk 5 above)
Comment (mdantonio) [reply src/environments/environment.prod.ts]: Yep, let's add an enable somewhere please 🙏

File: utility/config-env.ts (hunk 6 above)
Comment (mdantonio) [reply utility/config-env.ts]: It is, but on the container should be mapped from a variable defined on project_defaults/project_configuration/projectrc

Have a different name is possible, but the same name could be more consistent and easier. And by having the same name the above comment is then applied.

I.e. to define `multiLang: '${process.env.SOMETHING}'` in `utility/config-env.ts`, we would need

`SOMETHING: ${SOMETHING}` in `angular.yml`

We could have something consistent like:

```
multiLang: '${process.env.ENABLE_ANGULAR_MULTI_LANGUAGE}

.....

ENABLE_ANGULAR_MULTI_LANGUAGE: ${ENABLE_ANGULAR_MULTI_LANGUAGE}
```

Or eventually


```
multiLang: '${process.env.ENABLE_MULTI_LANGUAGE}

.....

ENABLE_MULTI_LANGUAGE: ${ENABLE_ANGULAR_MULTI_LANGUAGE}
```

if "angular" is redundant inside the container... but have the same name looks simpler to me


Once moved these considerations outside the container, identify the variable as "angular" only can avoid ambiguities

File: src/app/components/navbar/navbar.html (hunk 3 above)
Comment (giuseppetrotta) [reply src/app/components/navbar/navbar.html]: Back to this issue: what about the selected language as label? alternatively, indicate the selected language in the drop-down menu

Comment (giuseppetrotta) [reply src/app/components/navbar/navbar.html]: I was thinking of something like that

image

**Pull request merged**